### PR TITLE
Fix esbuild devserver

### PR DESCRIPTION
**Pull request by nicksnyder**
Our esbuild dev server is currently broken due to Elm in our codebase. 

```
[            web] ✘ [ERROR] No loader is configured for ".elm" files: client/web/src/search/results/components/compute/src/Main.elm
[            web]     client/web/src/notebooks/blocks/compute/NotebookComputeBlock.tsx:10:20:
[            web]       10 │ ...lm } from '../../../search/results/components/compute/src/Main....
[            web]          ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This PR fixes the devserver.

## Test plan

This doesn't change our product, just or dev server. Tested devserver locally.

## App preview:

- [Web](https://sg-web-ns-esbuild-devserver.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-jqcrcxhyud.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by burmudar** [comment]
I tried my hand in trying to at least get to the bottom of the elm error:
```
[            web] ✘ [ERROR]
[            web]     client/web/src/notebooks/blocks/compute/NotebookComputeBlock.tsx:10:20:
[            web]       10 │ ...lm } from '../../../search/results/components/compute/src/Main....
[            web]          ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
This is what I got so far after doing questionable things 👯 
```
client/web/src/notebooks/blocks/compute/NotebookComputeBlock.tsx:10:20: ERROR: [plugin: elm]
    at failureErrorWithLog (/Users/william/code/sourcegraph/node_modules/esbuild/lib/main.js:1624:15)
    at /Users/william/code/sourcegraph/node_modules/esbuild/lib/main.js:1266:28
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  errors: [
    {
      detail: undefined,
      id: '',
      location: [Object],
      notes: [],
      pluginName: 'elm',
      text: ''
    }
  ],
```
Not particular helpful, but it at least explains why we aren't getting additional output ... I think 🤔

**Comment by burmudar** [comment]
TL;DR: I was nerd sniped. Use aboslute paths instead of relative paths
I found the reason why we were getting that error. Which isn't even a elm error but I'd say a series of bad error reporting 😅.

`esbuild-plugin-elm` calls compileToStringSync from `node-elm-compiler`, which on any error throws [`throw 'Compilation failure.'`](https://sourcegraph.com/github.com/rtfeldman/node-elm-compiler/-/blob/src/index.ts?L181:5)). Since it assumed that the elm compiler would have outputted something about why compilation failed.

Going into `node-elm-compiler` and printing out the result of `compileSync` at https://sourcegraph.com/github.com/rtfeldman/node-elm-compiler/-/blob/src/index.ts?L174, we get
```
{"error":{"errno":-2,"code":"ENOENT","syscall":"spawnSync ./node_modules/.bin/elm","path":"./node_modules/.bin/elm","spawnargs":["make","/Users/william/code/sourcegraph/client/web/src/search/results/components/compute/src/Main.elm","--debug","--output","/var/folders/1r/5z42n9p52zv8rfp93gxc1vfr0000gn/T/2022715-47767-9g57lu.z9y6q.js"]},"status":null,"signal":null,"output":null,"pid":48407,"stdout":null,"stderr":null}
```
So some internal node failure is happening? Curiously if we look at the start of `index.ts` we'll find a function that converts err codes to legable errors ... and `ENOENT` means that the Elm executable was not found. So why is `node-elm-compiler` getting the wrong path ?

If we look at the executable it is trying to execute `./node_modules/.bin/elm`, and indeed with some strategic log statements the relative path `.` points to `sourcegraph/node_modules/esbuild-plugin-elm/` thus
 unlikely a relative path is going to work. It would be better to use a absolute path.

So to fix it, we have to tell `esbuild-plugin-elm` to resolve the `./node_modules/.bin/elm` to its absolute path. I've patched `esbuild-plugin-elm/index.js` in `sourcegraph/node_modules` with:
```
const getPathToElm = () => {
  if (fileExists('./node_modules/.bin/elm')) return path.resolve('./node_modules/.bin/elm')
  if (cmdExists('elm')) return 'elm'

  throw new Error('Could not find `elm` executable. You can install it with `yarn add elm` or `npm install elm`')
};
```
And it works :) I've also submitted a fix to `esbuild-plugin-elm` [here](https://github.com/phenax/esbuild-plugin-elm/pull/22) 
```
DEV_WEB_BUILDER=esbuild ./node_modules/.bin/gulp --color dev
[19:50:54] Using gulpfile ~/code/sourcegraph/gulpfile.js
[19:50:54] Starting 'dev'...
[19:50:54] Starting 'schema'...
[19:50:54] Starting 'graphQlSchema'...
[19:50:54] Starting 'graphQlOperations'...
[19:50:54] Starting 'cssModulesTypings'...
[19:50:55] Finished 'schema' after 799 ms
[19:50:57] Finished 'graphQlSchema' after 3.14 s
[19:50:58] Finished 'cssModulesTypings' after 3.23 s
[19:51:03] Finished 'graphQlOperations' after 8.84 s
[19:51:03] Starting 'esbuildDevelopmentProxy'...
[19:51:03] Starting 'watchSchema'...
[19:51:03] Starting 'watchGraphQlSchema'...
[19:51:03] Starting 'watchGraphQlOperations'...
[19:51:03] Starting 'watchCSSModulesTypings'...
🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥
ℹ  info      esbuild usage is experimental. See https://docs.sourcegraph.com/dev/background-information/web/build#esbuild.
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
▲ [WARNING] Comparison using the "!==" operator here is always true [equals-new-object]

    client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx:325:76:
      325 │         if (codeHostProblems.length === 0 && affiliatedRepos.codeHostErrors !== []) {
          ╵                                                                             ~~~

  Equality with a new object is always false in JavaScript because the equality operator tests
  object identity. You need to write code to compare the contents of the object instead. For
  example, use "Array.isArray(x) && x.length === 0" instead of "x === []" to test for an empty
  array.

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    node_modules/vscode-languageserver-types/lib/umd/main.js:3:24:
      3 │         var v = factory(require, exports);
        ╵                         ~~~~~~~

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    node_modules/monaco-editor/esm/vs/base/common/insane/insane.js:26:111:
      26 │ ...) { if (!n[i]) { if (!e[i]) { var c = "function" == typeof require && require; if (!f && c) return c(i, !0); if (u) return u...
         ╵                                                               ~~~~~~~

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    node_modules/monaco-editor/esm/vs/base/common/insane/insane.js:26:122:
      26 │ ...i]) { if (!e[i]) { var c = "function" == typeof require && require; if (!f && c) return c(i, !0); if (u) return u(i, !0); va...
         ╵                                                               ~~~~~~~

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    node_modules/monaco-editor/esm/vs/base/common/insane/insane.js:26:472:
      26 │ ... } return n[i].exports } for (var u = "function" == typeof require && require, i = 0; i < t.length; i++)o(t[i]); return o } ...
         ╵                                                               ~~~~~~~

● [DEBUG] Indirect calls to "require" will not be bundled [indirect-require]

    node_modules/monaco-editor/esm/vs/base/common/insane/insane.js:26:483:
      26 │ ...[i].exports } for (var u = "function" == typeof require && require, i = 0; i < t.length; i++)o(t[i]); return o } return r })...
         ╵                                                               ~~~~~~~

Success!

    Main ───> /var/folders/1r/5z42n9p52zv8rfp93gxc1vfr0000gn/T/2022715-36177-13woenr.9scr.js

4 warnings

  ui/assets/chunks/chunk-worker-AM63MEFH.js                               3.8mb ⚠️
  ui/assets/chunks/chunk-chunk-N55UPMA3.js                                3.7mb ⚠️
  ui/assets/chunks/chunk-chunk-E7QLJYU7.js                                3.0mb ⚠️
  ui/assets/scripts/app.js                                                1.4mb ⚠️
  ui/assets/chunks/chunk-chunk-UNONH5RK.js                                1.4mb ⚠️
  ui/assets/chunks/chunk-chunk-QTGGCM73.js                              920.5kb
  ui/assets/scripts/app.css                                             717.2kb
  ui/assets/chunks/chunk-chunk-62BSIC6E.js                              714.6kb
  ui/assets/chunks/chunk-chunk-AJMDW5PO.js                              576.3kb
  ui/assets/chunks/chunk-chunk-NXMH57C5.js                              519.1kb
  ui/assets/chunks/chunk-chunk-2MCZFMY5.js                              480.6kb
  ui/assets/chunks/chunk-chunk-ZEC3PUI2.js                              414.7kb
  ui/assets/chunks/chunk-RepositoryFileTreePage-BPVZMF3J.js             403.9kb
  ui/assets/chunks/chunk-chunk-TBBE6PFC.js                              391.5kb
  ui/assets/chunks/chunk-esm-RHXXMZPS.js                                349.2kb
  ui/assets/chunks/chunk-chunk-W34HXGUQ.js                              322.9kb
  ui/assets/chunks/chunk-jsonMode-VMRKX4G7.js                           292.3kb
  ui/assets/chunks/chunk-RepositoryFileTreePage-3TJABHSB.css            290.9kb
  ui/assets/chunks/chunk-SearchPage-4VKC2CML.js                         267.4kb
  ui/assets/chunks/chunk-NotebookPage-XHFS2KBM.css                      250.6kb
  ui/assets/chunks/chunk-chunk-PSWKFPRO.js                              245.5kb
  ui/assets/chunks/chunk-SearchPage-7YULUBZZ.css                        243.4kb
  ui/assets/chunks/chunk-chunk-WKHKC3X7.js                              234.2kb
  ui/assets/chunks/chunk-StreamingSearchResults-WQOXTRUM.css            226.5kb
  ui/assets/chunks/chunk-chunk-TRU5MEVQ.js                              221.3kb
  ui/assets/chunks/chunk-Kubernetes-RERVKKHQ.css                        217.5kb
  ui/assets/chunks/chunk-StackStorm-MDJBFBNQ.css                        217.5kb
  ui/assets/chunks/chunk-Stanford-ICS3XAP3.css                          217.5kb
  ui/assets/chunks/chunk-Temporal-UEESYAZ3.css                          217.5kb
  ui/assets/chunks/chunk-chakraui-DFIFMJZ7.css                          217.5kb
  ui/assets/chunks/chunk-Julia-OGVUQJOV.css                             217.5kb
  ui/assets/chunks/chunk-cncf-P67VEIXW.css                              217.5kb
  ui/assets/chunks/chunk-o3de-7S336FBU.css                              217.5kb
  ui/assets/chunks/chunk-SearchConsolePage-NFTIKM3Q.css                 215.6kb
  ui/assets/chunks/chunk-chunk-5PL5GLGJ.js                              181.7kb
  ui/assets/chunks/chunk-OrgSettingsArea-CQHBQZSA.css                   179.8kb
  ui/assets/chunks/chunk-chunk-VKGBIQY3.js                              179.3kb
  ui/assets/chunks/chunk-chunk-CN7UQ3UL.js                              176.1kb
  ui/assets/chunks/chunk-ExternalServicePage-BNHB4Y7V.css               172.3kb
  ui/assets/chunks/chunk-CodeMirrorQueryInput-4JW5O77L.css              171.0kb
  ui/assets/chunks/chunk-SiteAdminAddExternalServicesPage-ACCQ7AZR.css  169.4kb
  ...and 788 more output files...

⚡ Done in 4087ms
# esbuild: build took 4088ms
```
_Sorry for the wall of text_

**Comment by nicksnyder** [comment]
Awesome! I got it working by just providing the absolute path in the config so we don't have to wait for upstream.

**Review by valerybugakov — APPROVED** [review]
Awesome, thanks for looking into it, @burmudar and @nicksnyder! 🙌
